src/preact/amp-base-element.js

```diff
@@ -0,0 +1,132 @@
+  /**
+   * todo(kvchari): confirm this can be safely moved
+   * @override
+   */
+  mutatedAttributesCallback() {
+    if (this.container_) {
+      this.scheduleRender_();
+    }
+  }
+
+  /**
+   * todo(kvchari): confirm this can be safely moved
+   * @param {number} newHeight
+   * @override
+   */
+  attemptChangeHeight(newHeight) {
+    return super.attemptChangeHeight(newHeight).catch((e) => {
+      // It's okay to disable this lint rule since we check that the restricted
+      // method exists.
+
+      if (this.getOverflowElement && !this.getOverflowElement()) {
+        console./* OK */ warn(
+          '[overflow] element not found. Provide one to enable resizing to full contents.',
+          this.element
+        );
+      }
+      throw e;
+    });
+  }
+
+  /** @override */
+  getDefaultProps() {
+    return {
+      ...super.getDefaultProps(),
+      'onLoading': () => {
+        this.handleOnLoading();
+      },
+      'onLoad': () => {
+        this.handleOnLoad();
+      },
+      'onError': () => {
+        this.handleOnError();
+      },
+    };
+  }
+
+  /**
+   * Default handler for onLoad event
+   * Displays loader. Override to customize.
+   * @protected
+   */
+  handleOnLoad() {
+    this.toggleLoading?.(false);
+    this.toggleFallback?.(false);
+    this.togglePlaceholder?.(false);
+  }
+
+  /**
+   * Default handler for onLoading event
+   * Reveals loader. Override to customize.
+   * @protected
+   */
+  handleOnLoading() {
+    this.toggleLoading?.(true);
+  }
+
+  /**
+   * Default handler for onError event
+   * Displays Fallback / Placeholder. Override to customize.
+   * @protected
+   */
+  handleOnError() {
+    this.toggleLoading?.(false);
+    // If the content fails to load and there's a fallback element, display the fallback.
+    // Otherwise, continue displaying the placeholder.
+    if (this.getFallback?.()) {
+      this.toggleFallback?.(true);
+      this.togglePlaceholder?.(false);
+    } else {
+      this.togglePlaceholder?.(true);
+    }
+  }
+
+  /**
+   * Register an action for AMP documents to execute an API handler.
+   *
+   * This has no effect on Bento documents, since they lack an Actions system.
+   * Instead, they should use `(await element.getApi()).action()`
+   * @param {string} alias
+   * @param {function(API_TYPE, *):void} handler
+   * @param {ActionTrust_Enum} minTrust
+   * @protected
+   */
+  registerApiAction(alias, handler, minTrust = ActionTrust_Enum.DEFAULT) {
+    this.registerAction?.(
+      alias,
+      /** @param {*} invocation */
+      (invocation) => {
+        handler(this.api(), invocation);
+      },
+      minTrust
+    );
+  }
+}
+
+/**
+ * Changes the inheritance hierarchy of X such that it now extends from Y.
+ * If X previously inherited from Z, it now no longer does so.
+ * Dangerous, use sparingly!
+ * @param {*} X
+ * @param {*} Y
+ * @return {*}
+ */
+export function setSuperClass(X, Y) {
```

Comment:
```suggestion
export function setSuperClass(X, Y = AmpPreactBaseElement) {
```
if we default Y to `AmpPreactBaseElement` then we don't need to always import `AmpPreactBaseElement` into our amp classes, and we can always override it for the few cases where we need to change the prototype inheritance.

Reply:
I had a similar question that I was planning on asking later.
How many times are we calling setSuperClass with the same arguments? Ideally this is something that only needs to happen once.

Reply:
I agree that we should never call `setSuperClass` with the same arguments. Below are all the usages of `setSuperClass()`. I believe that `setSuperClass()` is never called twice with the exact same arguments. However, we call it 22 times with `AmpPreactBaseElement` and 6 times with `AmpVideoBaseElement` as the 2nd argument. (In all of these usages, `BaseElement` refers to different - bento - elements with the same generic name)

Re @dethstrobe's suggesion: I agree that it would look cleaner to maybe use a default for the 2nd parameter, but I also agree that this would further obfuscate the code make the more magical. Honestly my preference would be to keep it how it is (or make it not return as suggested [here](https://github.com/ampproject/amphtml/pull/37309#discussion_r779919449)) so we avoid magic. 


```js
class Amp__component_name_pascalcase__ extends setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpAccordion extends                     setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpBaseCarousel extends                  setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpDateCountdown extends                 setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpDateDisplay extends                   setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpEmbedlyCard extends                   setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpFacebookBase extends                  setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpFitText extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpIframe extends                        setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpInlineGallery extends                 setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpInstagram extends                     setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpLightbox extends                      setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpLightboxGallery extends               setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpMathml extends                        setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpRender extends                        setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSelector extends                      setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSidebar extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSocialShare extends                   setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpSoundcloud extends                    setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpStreamGallery extends                 setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpTimeago extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpTwitter extends                       setSuperClass(BaseElement, AmpPreactBaseElement) {...}
class AmpWordPressEmbed extends                setSuperClass(BaseElement, AmpPreactBaseElement) {...}

class AmpBrightcove extends                    setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpDailymotion extends                   setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpJwplayer extends                      setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpVideoIframe extends                   setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpVimeo extends                         setSuperClass(BaseElement, AmpVideoBaseElement) {...}
class AmpYoutube extends                       setSuperClass(BaseElement, AmpVideoBaseElement) {...}

class AmpEmbedlyKey extends                    setSuperClass(EmbedlyKeyBaseElement, AmpPreactBaseElement) {...}
class AmpFacebookComments extends              setSuperClass(CommentsBaseElement, AmpFacebookBase) {}
class AmpFacebookLike extends                  setSuperClass(LikeBaseElement, AmpFacebookBase) {}
class AmpFacebookPage extends                  setSuperClass(PageBaseElement, AmpFacebookBase) {}
class AmpInlineGalleryPagination extends       setSuperClass(PaginationBaseElement, AmpPreactBaseElement) {...}
class AmpInlineGalleryThumbnails extends       setSuperClass(ThumbnailsBaseElement, AmpPreactBaseElement) {...}
class AmpVideoBaseElement extends              setSuperClass(BentoVideoBaseElement, AmpPreactBaseElement) {...}
```